### PR TITLE
Add span-based Uri {Try}{Un}EscapeDataString overloads

### LINKED ISSUE #40603: [Proposal] Add new overload to Uri.EscapeDataString
Updated by @MihaZupan on 2024-01-15.

## Background and Motivation

In low allocation scenarios, it may be useful to support escaping URI data into an existing `Span<char>` buffer, rather than allocating a new `string`. 

As an example, see this [performance improvement](https://github.com/elastic/elasticsearch-net/issues/4951) suggestion for Elasticsearch.NET. The proposed change removes allocations on a hot path for query string building. It uses the existing `Uri.EscapeDataString` method which allocates a new string, only when the input contains characters that need encoding. This could be avoided if I were able to encode directly into the buffer rented from the `ArrayPool`.

## Proposed API

```c#
namespace System
{
    partial class Uri
    {
        // Existing
        public static string EscapeDataString(string stringToEscape);
        public static string UnescapeDataString(string stringToUnescape);

        // New
        public static string EscapeDataString(ReadOnlySpan<char> charsToEscape);
        public static bool TryEscapeDataString(ReadOnlySpan<char> charsToEscape, Span<char> destination, out int charsWritten);

        public static string UnescapeDataString(ReadOnlySpan<char> charsToUnescape);
        public static bool TryUnescapeDataString(ReadOnlySpan<char> charsToUnescape, Span<char> destination, out int charsWritten);
    }
}
```

## Usage Examples

The `foreach` loop from the linked issue could then be re-written along the lines of:

```c#
foreach (var key in nv.AllKeys)
{
    if (position != 1)
        bufferSpan[position++] = '&';

    // assume we are sure we have space in the destination
    if (Uri.TryEscapeDataString(key, bufferSpan.Slice(position), out var charsWritten))
        position += charsWritten;

    var value = nv[key];
    if (!value.IsNullOrEmpty())
    {
        bufferSpan[position++] = '=';
        if (Uri.TryEscapeDataString(value, bufferSpan.Slice(position), out charsWritten))
            position += charsWritten;
    }
}
```

In cases where the keys/values contain reserved characters, this avoids the temporary string allocations.

## Alternative Designs

1. 
    The existing method names contain `string`.
    Is it okay for the new span-based overloads to share the same name for better discoverability, or should we consider alternative names, e.g. `EscapeData` / `EscapeDataChars`?
1.
    The above design follows other `Span<T>` based overloads I have seen in common use and assumes that the operation could fail if the destination is too small. An alternative would be to throw in that case and instead apply the following signature (where the return value indicates the chars written):
    ```c#
    public static int EscapeDataString(ReadOnlySpan<char> stringToEscape, Span<char> destination);
    ```

### PR DESCRIPTION
**Pull request by MihaZupan**
Closes #40603

```c#
namespace System
{
    partial class Uri
    {
        public static string EscapeDataString(ReadOnlySpan<char> charsToEscape);
        public static bool TryEscapeDataString(ReadOnlySpan<char> charsToEscape, Span<char> destination, out int charsWritten);

        public static string UnescapeDataString(ReadOnlySpan<char> charsToUnescape);
        public static bool TryUnescapeDataString(ReadOnlySpan<char> charsToUnescape, Span<char> destination, out int charsWritten);
    }
}
```

The test coverage for these was not great, so I added a bunch more combinations. I also moved similar test cases from `System.Runtime.Tests` into the Uri project.

I updated `FormUrlEncodedContent` to also use these, so it's now allocating a lot less. For a random usage I pulled from [eShopOnWeb tests](https://github.com/dotnet-architecture/eShopOnWeb/blob/2fa4a3d3368381818c6191122d037c6a70b302f5/tests/FunctionalTests/Web/Controllers/AccountControllerSignIn.cs#L63-L69):
| Method             | Toolchain  | Mean     | Error   | Ratio | Allocated | Alloc Ratio |
|------------------- |----------- |---------:|--------:|------:|----------:|------------:|
| EShopOnWebLogin    | main       | 513.7 ns | 1.69 ns |  1.00 |    2080 B |        1.00 |
| EShopOnWebLogin    | pr         | 291.3 ns | 1.47 ns |  0.57 |     592 B |        0.28 |

Looking at something like YARP's [`EncodePath`](https://github.com/microsoft/reverse-proxy/blob/379671ef2afd01e55168929c33c6b5cbcd0b06b0/src/ReverseProxy/Forwarder/RequestUtilities.cs#L132) helper with ASP.NET Core's [test paths](https://github.com/dotnet/aspnetcore/blob/14868cc8741fda987431003ea3b06ac01e7fc3f1/src/Http/Http.Abstractions/perf/Microbenchmarks/PathStringBenchmark.cs#L10-L12), the difference is
| Method | Scenario             | Mean      | Error    | Ratio | Allocated | Alloc Ratio |
|------- |--------------------- |----------:|---------:|------:|----------:|------------:|
| Old    | LongTestPath         |  95.51 ns | 0.702 ns |  1.00 |     400 B |        1.00 |
| New    | LongTestPath         |  70.58 ns | 0.291 ns |  0.74 |     344 B |        0.86 |
|        |                      |           |          |       |           |             |
| Old    | LongT(...)rcent [24] | 337.00 ns | 2.713 ns |  1.00 |     392 B |        1.00 |
| New    | LongT(...)rcent [24] | 198.49 ns | 0.892 ns |  0.59 |     280 B |        0.71 |
|        |                      |           |          |       |           |             |
| Old    | TestPath             |  67.28 ns | 0.439 ns |  1.00 |     112 B |        1.00 |
| New    | TestPath             |  44.98 ns | 0.129 ns |  0.67 |      56 B |        0.50 |

### CODE REVIEW

File: src/libraries/System.Private.Uri/src/System/UriHelper.cs [hunk 1]
```diff
@@ -118,26 +118,79 @@ public static unsafe string SpanToLowerInvariantString(ReadOnlySpan<char> span)
...
+                // Nothing to escape, just copy the original chars.
+                charsToEscape.CopyTo(destination);
+                charsWritten = charsToEscape.Length;
+                return true;
+            }
+
+            var vsb = new ValueStringBuilder(stackalloc char[Uri.StackallocThreshold]);
+
+            // This may throw for very large inputs. ¯\_(ツ)_/¯
+            vsb.EnsureCapacity(charsToEscape.Length - indexOfFirstToEscape);
```
Comment (MihaZupan) [inline src/libraries/System.Private.Uri/src/System/UriHelper.cs:141]: @stephentoub should we care about cases like:
a) The input span is larger than max array length, so we can't rent the temporary buffer. But _technically_ the destination might be large enough to hold the entire result.
b) When escaping and the output grows, it grows beyond array max length, but lower than destination length, so it _technically_ could work (just not via `ValueStringBuilder`).
c) When escaping the output grows beyond int32

and/or should we care about which exception we end up throwing in such cases?

Realistically I hope no one would ever hit this.

File: src/libraries/System.Private.Uri/tests/FunctionalTests/UriEscapingTest.cs [hunk 2]
```diff
@@ -24,193 +24,304 @@ public class UriEscapingTest
...
-            string output = Uri.UnescapeDataString(Escape(input));
-            Assert.Equal(input, output);
+                // Destination too small
+                if (expectedOutput.Length > 0)
+                {
+                    Assert.False(Uri.TryEscapeDataString(input, destination.AsSpan(0, expectedOutput.Length - 1), out charsWritten));
+                    Assert.Equal(0, charsWritten);
+                }
+
+                // Overlapped source/destination
```
Comment (MihaZupan) [inline src/libraries/System.Private.Uri/tests/FunctionalTests/UriEscapingTest.cs:222]: Supporting overlapped source/destinations if "free" with the current implementation approach (via `ValueStringBuilder`).
Alternatively, we could detect and throw here if we wanted to.

File: src/libraries/System.Private.Uri/src/System/UriHelper.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.Uri/src/System/UriHelper.cs:141]: > should we care about

We shouldn't corrupt anything or the like, but if we throw an OutOfMemoryException or the like when close to the Array.MaxLength limit even if we're not technically over it, that's fine, at least for now. If we get feedback that it's an issue, we can tweak it later, but hopefully folks don't have 2B character Uris.

File: src/libraries/System.Private.Uri/src/System/UriExt.cs [hunk 3]
```diff
@@ -552,28 +552,99 @@ internal unsafe bool InternalIsWellFormedOriginalString()
...
+            int indexOfFirstToUnescape = charsToUnescape.IndexOf('%');
+            if (indexOfFirstToUnescape < 0)
+            {
+                // Nothing to unescape, just return the original value.
+                return backingString ?? charsToUnescape.ToString();
+            }
+
+            var vsb = new ValueStringBuilder(stackalloc char[StackallocThreshold]);
+
+            // This may throw for very large inputs. ¯\_(ツ)_/¯
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/UriExt.cs:586]: We can probably do without the emoticon :)

File: src/libraries/System.Private.Uri/src/System/UriExt.cs [hunk 4]
```diff
@@ -552,28 +552,99 @@ internal unsafe bool InternalIsWellFormedOriginalString()
...
+                }
 
-            int position = stringToUnescape.IndexOf('%');
-            if (position == -1)
-                return stringToUnescape;
+                charsWritten = 0;
+                return false;
+            }
 
             var vsb = new ValueStringBuilder(stackalloc char[StackallocThreshold]);
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/UriExt.cs:621]: Could we initialize the ValueStringBuilder with the destination buffer instead? Then at the end, if the VSB still contains the original span, we can avoid copying it, and if it doesn't, it presumably means the content was too big and we can fail?

Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/UriExt.cs:621]: For the unescape path, most likely yes. I'll test how much of a difference it'd make.

For the escape path, we'd need a bit more code to handle overlapping input/output buffers. I'll measure it.

Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/UriExt.cs]: Did it in 16f70f42a7fca46358acf4e92f24f415d4234a93, doesn't seem too bad.
It's slightly faster now.

File: src/libraries/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs [hunk 5]
```diff
@@ -28,30 +28,61 @@ private static byte[] GetContentByteArray(IEnumerable<KeyValuePair<string?, stri
...
 
-                builder.Append(Encode(pair.Key));
+                Encode(ref builder, pair.Key);
                 builder.Append('=');
-                builder.Append(Encode(pair.Value));
+                Encode(ref builder, pair.Value);
             }
 
-            return HttpRuleParser.DefaultHttpEncoding.GetBytes(builder.ToString());
+            // We know the encoded length because the input is all ASCII.
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs:45]: How do we know the input is ASCII?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs (hunk 5 above)
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs:45]:
```suggestion
            // EscapeDataString will always return an ASCII string and DefaultHttpEncoding is Latin1,
            // so we know the output byte size will be the same as the builder length.
```

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs]: The input is a bunch of concatenated `Uri.EscapeDataString` results, which are always ASCII, and the encoding is Latin1.
I clarified it in the comment now.

**Pull request merged**